File: Doubly_Linked_Lists/Java/List.java
/*	List
	A simple doubly linked list.
	Written by Bailey Darling
		September 2016

	List has a front, back, and current (curr)

	index indicates where curr is.
	if index is -1, its undefined
	0 is front
	length-1 is back

	clear:			clears the list
	moveFront:		moves curr to front
	moveBack:		moves curr to back
	moveNext:		moves curr forward one
	movePrev:		moves curr back one

	prepend:		appends to front of the list
	append:			appends to back of the list
	insertBefore:	appends before curr
	insertAfter:	appends after curr

	delete:			deletes current
	deleteFront:	deletes the front node
	deleteBack:		deletes the rear node

	length()		returns length
	index()			returns index value
	front()			returns front data
	back()			returns back data
	geth()			returns curr's data

	equals			returns true if two lists are identical
	isEmpty			returns true if empty

	Copy 			copies this into a new list
	toString		overloads standard toString


*/

class List{

	private class Node{
		//the node
		int data;
		Node Next;
		Node Prev;

		Node(int data){
			this.data=data;
			Next = null;
			Prev = null;
		}

		public String toString(){
			return String.valueOf(data);

		}

		public boolean equals(Node x){
			//if equal to current node, returns true
			return ((this.data)==(x.data));
		}
	}

	private Node front; Node back; Node curr;
	private int length; int index;

	List(){
		front = back = curr= null;
		length = 0;
		index = -1;
	}

	//--------------Manipulation Procedures---------------------//
	// clear : resets the list to empty

	void clear(){
		front = back = curr = null;
		index = -1;
		length = 0;

	}

	//moveFront
	// if the list is not empty, moves curr to the front

	void moveFront(){
		if(!isEmpty()){
			index=0;
			curr=front;
		}
	}

	//moveBack
	// if list is not empty, moves curr to the back
	void moveBack(){
		if(!isEmpty()){
			index=length-1;
			curr=back;
		}

	}

	//moveNext
	// if list is not empty, moves curr to next node
	// if there is no next node, curr is undefined and index is -1
	void moveNext(){
		if(!isEmpty()){
			if(index==length-1){
				index= -1;
				curr = null;
			}else{
				index++;
				curr=curr.Next;
			}
		}

	}

	//movePrev
	//if list is not empty, moves curr to previous node
	//if there is no previous node, curr is null and index is -1

	void movePrev(){
		if(!isEmpty()){
			if(index==0){
				index=-1;
				curr=null;

			}else{
				index--;
				curr=curr.Prev;
			}
		}
	}

	// ------------- insertions -----------------//

	//appends to the front of the list
	void prepend(int data){
		Node R = new Node(data);
		if(isEmpty()){
			front=back=R;
		}else{
			R.Next=front;
			front.Prev=R;
			front=R;
		}
		length++;
	}

	//appends to the back of the list
	void append(int data){
		Node R= new Node(data);
		if(this.isEmpty()){
			front=back=R;
		}else{
			back.Next=R;
			R.Prev=back;
			back=R;
		}
		length++;
	}

	void insertBefore(int data){
		if(length >=0 && index >=0){
			Node R = new Node(data);
			if(this.isEmpty()){
				front=back=R;
			}else if(index==-1 ||curr.Prev ==null){
				prepend(data);
			}else{
				curr.Prev.Next=R;
				curr.Prev=R;
				R.Prev=curr.Prev;
				R.Next=curr;
				index++;
			}
			length++;
		}

	}

	void insertAfter(int data){
		if(length>=0 && index<=length-1){
			Node R = new Node(data);
			if(this.isEmpty()){
				front=back=R;
			}else if(index==-1 || curr.Next==null){
				append(data);
			}else{
				curr.Next.Prev=R;
				R.Next=curr.Next;
				R.Prev=curr;
				curr.Next=R;
			}
		}
		length++;
	}

	//-------------deletions------------------//
	//Deletes the front element
	void deleteFront(){
		front=front.Next;
		length--;
	}
	//Deletes the rear element
	void deleteBack(){
		back=back.Prev;
		length--;
	}
	//Deletes the cursor element
	void delete(){
		if(length>0 &&index>=0){
			curr.Prev.Next= curr.Next;
			curr.Next.Prev=curr.Prev;
			curr=null;
			index=-1;
			length--;
		}
	}


	//---------------Access functions -----------------------//

	//returns the length of the list
	int length(){
		return length;
	}

	//returns the index
	// returns -1 if cursor is undefined
	int index(){
		return index;
	}

	//returns the front element
	int front(){
		if(this.isEmpty()){
			throw new RuntimeException(
				"front() cannot be called on an empty list");
		}
		return front.data;
	}
	
	//returns rear element
	int back(){
		if(this.isEmpty()){
			throw new RuntimeException(
				"back() cannot be called on an empty list");	
		}
		return back.data;
	}

	//returns the curr element
	int get(){
		if(this.isEmpty()){
			throw new RuntimeException(
				"get() cannot be called on an empty list");	
		}
		return curr.data;
	}


	//Ignores cursors
	//returns true if lists are identical
	boolean equals(List L){
		Node N= this.front;
		Node K= L.front;

		boolean equal = (this.length()== L.length());

		while(N!= null){
			if(!equal) return false;

			equal = N.equals(K);
			N=N.Next;
			K=K.Next;
		}
		return equal;
	}

	boolean isEmpty(){
		return length==0;
	}

	public String toString(){
		StringBuffer sb = new StringBuffer();
		Node N= front;
		while(N!=null){
			sb.append(" ");
			sb.append(N.toString());
			N=N.Next;
		}
		return new String(sb);
	}

// Copy
// Copies the list

	List copy(){
		List R= new List(); // the new list
		Node N= this.front; //the start of the list to be copied

		while(N!=null){
			R.append(N.data);
			N=N.Next;
		}
		R.length = this.length;
		return R;
	}

}

File: Doubly_Linked_Lists/C/List.c
/*	List
	A simple doubly linked list.
	Written by Bailey Darling
		September 2016

	List has a front, back, and current (curr)

	index indicates where curr is.
	if index is -1, its undefined
	0 is front
	length-1 is back

	length()		returns length
	index()			returns index value
	front()			returns front data
	back()			returns back data
	get()			returns curr's data

	clear:			clears the list
	moveFront:		moves curr to front
	moveBack:		moves curr to back
	moveNext:		moves curr forward one
	movePrev:		moves curr back one

	prepend:		appends to front of the list
	append:			appends to back of the list
	insertBefore:	appends before curr
	insertAfter:	appends after curr

	delete:			deletes current
	deleteFront:	deletes the front node
	deleteBack:		deletes the rear node

	equals			returns true if two lists are identical
	isEmpty			returns true if empty

	Copy 			copies this into a new list
	toString		overloads standard toString


*/

	#include <stdio.h>
	#include <stdlib.h>
	#include "List.h"

////////////////////////Structs///////////////////////////////////

	//Node struct
typedef struct NodeObj{
		int data;
		struct NodeObj* next;		//pointers for connecting nodes
		struct NodeObj* prev;

	}NodeObj;

	typedef NodeObj* Node;			//Initialize

typedef struct ListObj{
	Node front;
	Node back;
	Node curr;
	int length;
	int index;
}ListObj;

Node newNode(int data){
	Node N= malloc(sizeof(NodeObj));
	N->data = data;
	N->next = NULL;
	N->prev = NULL;

		return(N);
	}

	void freeNode(Node* pN){
		if(pN!=NULL && *pN != NULL){
			free(*pN);
			*pN= NULL;
		}
	}

//Creates a new List

List newList(void){
	List L;
	L = malloc(sizeof(ListObj));
	L->front= L->back = NULL;
	L->length = 0;
	L->index = -1;
	L->curr = NULL;

	return L; 
}

// Helper function that Frees a List //

void freeList(List* L){
	if(L!= NULL && *L!=NULL){
		while(length(*L)!=0){
			deleteFront(*L);
		}
		clear(*L);
		free(*L);
		*L= NULL;
	}
}


// --------------------- Access Functions -----------------------//

int length(List L){
	if(L==NULL){
		printf("ERROR: Cannot get length for a NULL reference\n");
		exit(1);
	}
	return (L->length);
}

int index(List L){
	if(L==NULL){
		printf("ERROR: Cannot get index of a NULL reference\n");
		exit(1);
	}
	return(L->index);
}

int front(List L){
	if(L==NULL){
		printf("ERROR: Cannot get front of a NULL reference\n");
		exit(1);
	}
	return(L->front->data);
}

int back(List L){
	if(L==NULL){
		printf("ERROR: Cannot get back of a NULL reference\n");
		exit(1);
	}
	return(L->back->data);
}

int get(List L){
	if(L==NULL){
		printf("ERROR: Cannot get from a NULL reference\n");
		exit(1);
	}else if(L->index == -1){
		printf("ERROR: Cursor is off the list\n");
		return -1;
	}else{
		return L->curr->data;
	}
}

int equals(List A, List B){
	//Returns true only if A and B are equal

	if(A == NULL || B == NULL){
		printf("ERROR: called equals() on a NULL reference\n");
	}
	Node N = NULL;
	Node M = NULL;
	int flag = 0;
	flag = (A->length == B->length);

	N = A->front;
	M = B->front;

	while(flag && N!=NULL){
		flag = N->data == M->data;
		N = N->next;
		M = M->next;
	}
	return flag;
}

// --------------------- Manipulation Functions ------------------------//

void clear(List L){				// helper fn, clears a list.
	if(L->front == NULL) return;

	Node tmp1 = L->front;
	Node tmp2;

	while(tmp1!= NULL){
		tmp2= tmp1->next;
		free(tmp1);
		tmp1=tmp2;
	}

	L->front = L->back = L->curr = NULL;
	L->length = 0;
	L->index = -1;
}

void moveFront(List L){				// curr = front
	if(L->length != 0){
		L->curr = L->front;
		L->index = 0;
	}
}

void moveBack(List L){				// curr = back
	if(L->length != 0){
		L->curr = L->back;
		L->index = L->length-1;
	}
}

void movePrev(List L){				// decrements curr
	if(index(L)== -1) return;
	if(length(L) >= 0){
		if(length(L)==0){
			L->curr = NULL;
			L->index= -1;
			return;
		}
		L->curr = L->curr->prev;
		L->index--;
	}
}

void moveNext(List L){				//increments curr
	if(L->length != 0){
		if(L->index == length(L)-1){
			L->curr = NULL;
			L->index = -1;
		}
		else{
			L->curr = L->curr->next;
			L->index++;
		}
	}
}

void prepend(List L, int data){		//Appends to the beginning of the list
	Node N = newNode(data);
	if(L->length==0){
		L->front= L->back = N;
	}else{
		N->next= L->front;
		L->front->prev = N;
		L->front= N;
	}
	L->length++;
}

void append(List L, int data){		//Appends to the end of the list
	Node N = newNode(data);
	if(L->length==0){
		L->front=L->back=N;
	}else{
		N->prev = L->back;
		L->back->next = N;
		L->back = N;
	}
	L->length++;
}

void insertBefore(List L, int data){ //inserts before curr
	Node N= newNode(data);
	if(L->length==0){
		L->front=L->back=N;
	}else if(L->index == -1 || L->curr->prev == NULL){
		printf("ERROR: Cursor undefined\n");
		return;
	}else{
		L->curr->prev->next = N;
		N->next = L->curr;
		N->prev = L->curr->prev;
		L->curr->prev = N;
		L->index++;
	}
	L->length++;
}

void insertAfter(List L, int data){
	Node N = newNode(data);
	if(L->length==0){
		L->front=L->back=N;
	}else if(L->index == -1 || L->curr->prev == NULL){
		printf("ERROR: Cursor undefined\n");
		return;
	}else{
		L->curr->next->prev = N;
		N->next = L->curr->next;
		N->prev = L->curr;
		L->curr->next = N;
	}
	L->length++;
}

void deleteFront(List L){


    if(L->length==0){
        printf("Error: cannot call a delete function on an empty List");
        exit(1);
    }
    Node N = NULL;
    N=L->front;
    if(L->length>1){
        L->front = L->front->next;
        L->front->prev = NULL;
    }else{
        L->front=L->back= L->curr =NULL;
    }
    L->length--;
    freeNode(&N);
}



void deleteBack(List L){
    Node N = NULL;

    if(L->length==0){
        printf("Error: cannot call a delete function on an empty List");
        exit(1);
    }
    N= L->back;
    if(L->length>1){

        L->back = L->back->prev;
        L->back->next = NULL;
    }else{
        L->front=L->back=L->curr=NULL;
        L->index=-1;
    }
    L->length--;
    freeNode(&N);

}

void delete(List L){
    //deletes curr
    Node N = NULL;
    if(L->length==0){
        printf("Error: cannot call a delete function on an empty List");
        exit(1);
    }
    N= L->curr;
    if(L->length>1){
        L->curr->prev->next = L->curr->next;
        L->curr->next->prev = L->curr->prev;
        L->length--;
    }else{
        L->front=L->back=NULL;
    }
        L->curr = NULL;
        L->index = -1;

        freeNode(&N);

}

//--------------------------Other ----------------------------//
List copyList(List L){
    List K = newList();
    Node N;
    N = L->front;

    while(N != NULL){
        append(K, N->data);
        N = N->next;
    }

    return(K);

}

int isEmpty(List L){
    return(L->length==0);
}

void printList(FILE* out, List L){
    //Node T;
    for(moveFront(L); index(L)!= -1; moveNext(L)){
        fprintf(out, "%d ", get(L));
    }
}

































